File: 2021/deno/day-1.ts
const input = await Deno.readTextFile("./input/day-1.in");

const sum = (a: number, b: number) => a + b;

const data = input.split("\n").map(Number);

/**
 * Part One
 */

const deltas = data
  .reduce((prev: boolean[], curr, index, src) => {
    if (index === 0) return prev;
    return [...prev, curr > src[index - 1]];
  }, [])
  .filter(Boolean).length;

console.log("Part One:", deltas);

/**
 * Part Two
 */

const windowDeltas = data
  .reduce((prev: boolean[], _, index, src) => {
    if (index < 3) return prev;

    return [...prev, src[index] > src[index - 3]];
  }, [])
  .filter(Boolean).length;

console.log("Part One:", windowDeltas);

File: 2021/swift/day-1.swift
import Foundation

let filename = "../input/day-1.in"

func main() {
    do {
        let contents = try String(contentsOfFile: filename)

        let data = contents.split(separator: "\n").map {
            (val: String.SubSequence) -> Int in
            Int(val)!
        }

        func solver(size: Int, list: [Int]) -> Int {
            let slider: (Int, (Int, Int)) -> Int = {
                acc, curr -> Int in
                let (index, value) = curr

                if index <= size {
                    return acc
                }

                return acc + (value > list[index - size] ? 1 : 0)
            }

            return list.enumerated().reduce(0, slider)
        }

        print("Part one: \(solver(size: 1, list: data))")

        print("Part two: \(solver(size: 3, list: data))")
    } catch {
        print("Failed to read file")
    }
}

main()
